### PR TITLE
Adding scheduled time to all the things

### PR DESCRIPTION
**Pull request by pauljamescleary**
Add scheduled time field to prepare for scheduled batch changes.

* `create_batch_change_test.py` - add a test that ensures posting and retrieving using the scheduled time field works as expected
* `api/src/main/resources/application.conf` - fix an issue where the task repo wasn't loaded on `reStart`
* `BatchChangeProtocol.scala` - add scheduledTime field to `BatchChangeInput`
* `BatchChangeService.scala` - modify the `buildResonse` method, it does some logic to produce the resulting `BatchChange` entity, and we need to ensure that the `scheduledTime` field propagates through that logic
* `BatchChangeJsonProtocol.scala` - make sure that `BatchChangeInputSerializer` takes in `scheduledTime`, make sure that `BatchChangeSerializer` converts `scheduledTime` to json
* `BatchChange.scala` - add `scheduledTime` field
* `BatchChangeSummary.scala` - add `scheduledTime` field
* `V3.18__ScheduledChange.sql` - add a `scheduled_time` field to the `batch_change` table; add an index on `scheduled_time` as well to support querying by scheduled time
* `MySqlBatchChangeRepository` - make sure that save and get methods support `scheduledTime`

### CODE REVIEW

File: modules/api/src/main/scala/vinyldns/api/route/BatchChangeJsonProtocol.scala [hunk 1]
```diff
@@ -57,7 +58,8 @@ trait BatchChangeJsonProtocol extends JsonValidation {
       (
         (js \ "comments").optional[String],
         changeList,
-        (js \ "ownerGroupId").optional[String]
+        (js \ "ownerGroupId").optional[String],
+        (js \ "scheduledTime").optional[DateTime]
```
Comment (mitruly) [inline modules/api/src/main/scala/vinyldns/api/route/BatchChangeJsonProtocol.scala]: It may be worthwhile to providing the acceptable format (`YYYY-MM-DDTHH:MM:SSZ`) for this error message; unsupported value just yields the following error message (using input `2019-07-22T14:16:28.360Z`):
```
    "errors": [
        "Invalid date format 2019-07-22T14:16:28.360Z"
    ]
```

Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/route/BatchChangeJsonProtocol.scala]: Yea, that is a good idea, our invalid date message doesn't tell you how it should be done

File: modules/api/functional_test/live_tests/batch/create_batch_change_test.py [hunk 2]
```diff
@@ -269,6 +270,32 @@ def test_create_batch_change_with_adds_success(shared_zone_test_context):
...
+        "scheduledTime": dt
+    }
+
+    to_delete = []
+    try:
+        result = client.create_batch_change(batch_change_input, status=202)
+        completed_batch = client.wait_until_batch_change_completed(result)
+        record_set_list = [(change['zoneId'], change['recordSetId']) for change in completed_batch['changes']]
+        to_delete = set(record_set_list)
+        assert_that(completed_batch['scheduledTime'], dt)
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/batch/create_batch_change_test.py]: can you add an assert to another test to confirm it's not appearing as a field when empty? dont want people seeing this before its fully implemented

File: modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala [hunk 3]
```diff
@@ -58,7 +58,8 @@ object VinylDateParser {
     format.dateFormat
       .parse(s)
       .map(_.getTime)
-      .getOrElse(throw new MappingException(s"Invalid date format $s"))
+      .getOrElse(throw new MappingException(
+        s"Invalid date format $s; provide the date format as YYYY-MM-DDTHH:MM:SSZ"))
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala]: thats how we take in dates? how do we know what time zone they are talking about?

File: modules/api/src/main/resources/application.conf [hunk 4]
```diff
@@ -54,6 +54,9 @@ vinyldns {
       }
       record-set {
 
+      }
+      task {
```
Comment (rebstar6) [inline modules/api/src/main/resources/application.conf]: why are we putting the task repo in the api?

File: modules/mysql/src/main/resources/db/migration/V3.18__ScheduledChange.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CREATE SCHEMA IF NOT EXISTS ${dbName};
+
+USE ${dbName};
+
+ALTER TABLE batch_change ADD scheduled_time DATETIME NULL;
+
+CREATE INDEX scheduled_time_index ON batch_change (scheduled_time);
```
Comment (rebstar6) [inline modules/mysql/src/main/resources/db/migration/V3.18__ScheduledChange.sql]: not 100% certain we need this index. things should only become scheduled if the approval status is NeedsApproval, so youre only ever filtering on that smaller set if youre doing it right

**Comment by rebstar6** [comment]
sry merge conflicted you. you need to actually rebase and change `mapN(BatchChangeInput(_, _, _))` in the BatchChangeJsonProtocol

File: modules/mysql/src/main/resources/db/migration/V3.18__ScheduledChange.sql (hunk 5 above)
Comment (pauljamescleary) [reply modules/mysql/src/main/resources/db/migration/V3.18__ScheduledChange.sql]: Yea, I will drop the index, as this changed recently.

Instead of providing an endpoint that will allow one to query for elapsed changes, they will "get all in manual review that are open" and further filter for those that have a scheduledTime

File: modules/api/src/main/resources/application.conf (hunk 4 above)
Comment (pauljamescleary) [reply modules/api/src/main/resources/application.conf]: I was unable to do `reStart` without it

File: modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala (hunk 3 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala]: good point, need the timezone, should be in `ISO8601` format.  I will update the datetime parser and add a unit test

File: modules/api/functional_test/live_tests/batch/create_batch_change_test.py (hunk 2 above)
Comment (pauljamescleary) [reply modules/api/functional_test/live_tests/batch/create_batch_change_test.py]: Done

File: modules/api/src/main/resources/application.conf (hunk 4 above)
Comment (rebstar6) [reply modules/api/src/main/resources/application.conf]: thats...weird

File: modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala (hunk 3 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala]: @rebstar6 so "Z" means "Zulu time", which indicates UTC.  Basically we ONLY support second level resolution in UTC, so this is correct.

Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/route/JsonValidation.scala]: go figure. ok works for me

File: modules/api/src/main/resources/application.conf (hunk 4 above)
Comment (rebstar6) [reply modules/api/src/main/resources/application.conf]: just in this branch? im not seeing this issue in master
